vm/src/pyobjectrc.rs

```diff
@@ -255,37 +516,69 @@ impl PyObjectRef {
 
 impl PyObject {
     #[inline]
-    fn with_pyobjectref<R>(&self, f: impl FnOnce(&PyObjectRef) -> R) -> R {
-        // SAFETY: we don't use obj after the real lifetime (self) goes out of scope, we wrap it
-        //         in a ManuallyDrop to prevent double free, and we only pass it by reference
-        let obj = unsafe { ManuallyDrop::new(PyObjectRef::from_raw(self)) };
-        f(&obj)
+    fn weakreflist(&self) -> Option<&WeakRefList> {
+        Some(&self.0.weaklist)
     }
 
-    pub fn downgrade(&self) -> PyObjectWeak {
-        PyObjectWeak {
-            weak: self.with_pyobjectref(|obj| PyRc::downgrade(&obj.rc)),
-        }
+    pub(crate) fn downgrade_with_typ_opt(
+        &self,
+        callback: Option<PyObjectRef>,
+        typ: PyTypeRef,
+    ) -> Option<PyObjectWeak> {
+        self.weakreflist()
+            .map(|wr| wr.add(self, typ, callback, None))
+    }
+
+    pub(crate) fn downgrade_with_typ(
+        &self,
+        callback: Option<PyObjectRef>,
+        typ: PyTypeRef,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyObjectWeak> {
+        let dict = if typ
```

Comment:
I am confused here. A dict is created when downgrading?

Reply:
Yeah, for the `weakref` object - if it's a subclass it needs to have a dict for the user code to store attributes and stuff on, like any user subclass